### PR TITLE
Adds a new TransactionRepository with support for snapshots and concurrent writers

### PR DESCRIPTION
**Pull request by gingerhendrix**
Adds support for https://github.com/textmaster/snapshot_aggregate_root

## Snapshots

Snapshot support provided through: 
* create_snapshot - saves a snapshot to the event stream
* last_stream_snapshot - finds the most recent snapshot in the event stream

## Concurrency

In order to guarantee consistency, a `with_lock` method is provided to provide exclusive access to the event stream for writing.  In addition, in order to provide atomic writes, all the events written during the transaction (including snapshots) are written as a single document to mongodb.

## Misc

* Pagination can be disabled for this repository by supplying `count: nil` to the read methods (snapshots remove the need for pagination)

### CODE REVIEW

File: lib/rails_event_store_mongoid/lock.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+    def with_lock(stream)
+      begin
+        lock = self.class.create(_id: stream, ts: Time.now.utc)
+      rescue ::Mongo::Error::OperationFailure
+        raise CannotObtainLock
+      end
+      yield if block_given?
+    ensure
+      lock.delete if lock.present?
```
Comment (gottfrois) [inline lib/rails_event_store_mongoid/lock.rb:20]: tell don't ask, move the condition in the delete method

File: lib/rails_event_store_mongoid/locker.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+require 'rails_event_store_mongoid/lock'
+
+module RailsEventStoreMongoid
+  class Locker
+    def initialize(adapter: ::RailsEventStoreMongoid::Lock.new)
+      @adapter = adapter
```
Comment (gottfrois) [inline lib/rails_event_store_mongoid/locker.rb]: add an `attr_reader`

File: lib/rails_event_store_mongoid/locker.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    end
+
+    def with_lock(stream, &block)
+      begin
+        retry_count = 0
+        @adapter.with_lock(stream, &block)
+      rescue CannotObtainLock
+        retry_count += 1
+        if retry_count < 5
+          sleep(0.5)
```
Comment (gottfrois) [inline lib/rails_event_store_mongoid/locker.rb]: configurable? 0.1 or even no sleep should be fine i think

File: lib/rails_event_store_mongoid/transaction.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        tx = last_transaction(stream: stream)
+        return nil unless tx.present?
+        tx.last_event
+      end
+    end
+
+    index({ 'events.event_id': 1 }, { unique: true })
+    index(stream: 1, ts: 1)
+    # What is the index I want?
+    # index({ stream: 1, ts: 1}, { partialFilterExpress: { 'events.snapshot': true } })
```
Comment (gottfrois) [inline lib/rails_event_store_mongoid/transaction.rb]: sup?

File: lib/rails_event_store_mongoid/transaction_repository.rb [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        event_type: event.class,
+        data:       event.data,
+        meta:       event.metadata,
+      }
+    end
+
+    def build_snapshot_model(event)
+      build_event_model(event).merge({
+        snapshot: true,
+      })
```
Comment (gottfrois) [inline lib/rails_event_store_mongoid/transaction_repository.rb]: remove useless `{}`

File: lib/rails_event_store_mongoid/transaction_repository.rb [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    end
+
+    def read_forwards(adapter, start_event_id, count = 0)
+      stream = adapter
+      unless start_event_id.equal?(:head)
+        starting_tx = stream.with_event(start_event_id).first
+        starting = starting_tx.events.after(start_event_id)
+        stream = stream.where(:ts.gt => starting_tx.ts)
+      end
+      #TODO: This is count of transactions not count of events
```
Comment (gottfrois) [inline lib/rails_event_store_mongoid/transaction_repository.rb]: sup?

File: lib/rails_event_store_mongoid/transaction_repository.rb [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+      stream = stream.limit(count) if count > 0
+
+      (starting || []) + stream.asc(:ts).map(&method(:build_event_entities)).flatten(1)
```
Comment (gottfrois) [inline lib/rails_event_store_mongoid/transaction_repository.rb]: `Array(starting)`

**Comment by gingerhendrix** [comment]
Superseded by https://github.com/gingerhendrix/transaction_event_store_mongoid

**Pull request closed without merging**